### ISSUE TITLE
BUG: Passing float32 to datetime.timedelta throws TypeError in numpy 2.0.0

### ISSUE THREAD

**Issue by johncoxon** [numpy/numpy#26804]
### Describe the issue:

`datetime.timedelta` throws a TypeError when you pass an `np.float32` to it in numpy 2.0.0, despite saying in the documentation, "Arguments may be integers or floats, and may be positive or negative." It is fine in numpy 1.26.4 with the same versions of Python and python-dateutil (tested in 2.9.0).

### Reproduce the code example:

```python
import numpy as np
from datetime import timedelta
timedelta(seconds=np.float32(60))
```

### Error message:

```shell
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported type for timedelta seconds component: numpy.float32
```

### Python and NumPy Versions:

2.0.0
3.12.4 | packaged by conda-forge | (main, Jun 17 2024, 10:13:44) [Clang 16.0.6 ]

### Runtime Environment:

[{'numpy_version': '2.0.0',
  'python': '3.12.4 | packaged by conda-forge | (main, Jun 17 2024, 10:13:44) '
            '[Clang 16.0.6 ]',
  'uname': uname_result(system='Darwin', node='Yavin.local', release='23.5.0', version='Darwin Kernel Version 23.5.0: Wed May  1 20:16:51 PDT 2024; root:xnu-10063.121.3~5/RELEASE_ARM64_T8103', machine='arm64')},
 {'simd_extensions': {'baseline': ['NEON', 'NEON_FP16', 'NEON_VFPV4', 'ASIMD'],
                      'found': ['ASIMDHP'],
                      'not_found': ['ASIMDFHM']}},
 {'architecture': 'VORTEX',
  'filepath': '/opt/homebrew/Caskroom/mambaforge/base/envs/numpy/lib/libopenblas.0.dylib',
  'internal_api': 'openblas',
  'num_threads': 8,
  'prefix': 'libopenblas',
  'threading_layer': 'openmp',
  'user_api': 'blas',
  'version': '0.3.27'},
 {'filepath': '/opt/homebrew/Caskroom/mambaforge/base/envs/numpy/lib/libomp.dylib',
  'internal_api': 'openmp',
  'num_threads': 8,
  'prefix': 'libomp',
  'user_api': 'openmp',
  'version': None}]

### Context for the issue:

It's an incompatibility between numpy and Python itself; passing numpy floats to core Python functions that accept floats should always work.

**Comment by seberg** [comment]
Not a regression, can reproduce in NumPy 1.26.4.  Maybe you updated Python at the same time?

**Comment by johncoxon** [comment]
Apologies: I did create identical conda environments and only changed the NumPy version, but I didn't dig deep enough in my code.

It's not that the behaviour has changed in timedelta. What's changed in NumPy 2 is that the `np.round(x)` call I'm making returns a float32 instead of a float64 (x is a float32 in both situations), and timedelta accepts a float64 fine but not a float32. It probably makes sense that `np.round` returns a float32 in this situation now, so that should probably not be changed. Is it worth logging this as a bug with `timedelta` on the python/cpython repo instead?

**Comment by WarrenWeckesser** [comment]
I don't think this is a bug in `timedelta`.  Its docstring says "Arguments may be integers or floats", and presumably that means *Python* integers or floats.  `np.float64` happens to be a subclass of Python's `float`, so it is accepted.  This is not true for `np.float32` or `np.longdouble`.  You also can't pass objects such as `decimal.Decimal`, `fractions.Fraction`, `mpmath.mp.mpf`, etc.

**Comment by seberg** [comment]
Agreed, closing for now also.  The one thing that Python could do, is support integer objects which support `__index__()`, because index (should) promise to be a real integer, but since this also accepts floats, things are just tricky, probably.

FWIW, `np.round` also shouldn't have changed behavior, but you probably have some trivial math like `x + 3` or so.  And for that NumPy *did* indeed change behavior: before it would have upcast to float64, no it sticks to the explicit `float32` and effectively ignores the Python int or float.

**Comment by johncoxon** [comment]
Yes, I think it's the latter part that's happened here. Thanks!
